cupy/_core/_routines_sorting.pyx

```diff
@@ -231,21 +231,67 @@ cdef _ndarray_base _ndarray_argpartition(self, kth, axis):
     ndim = data._shape.size()
     _axis = internal._normalize_axis_index(_axis, ndim)
 
+    if axis == ndim - 1:
+        data = self
+    else:
+        data = _manipulation.rollaxis(self, axis, ndim).copy()
+
     length = data._shape[_axis]
+
+    if length == 0:
+        return cupy.empty((0,), dtype=cupy.int64)
+
     if isinstance(kth, int):
         kth = kth,
+    max_k = 0
     for k in kth:
         if k < 0:
             k += length
         if not (0 <= k < length):
             raise ValueError('kth(={}) out of bounds {}'.format(k, length))
+        if max_k < k:
+            max_k = k
 
-    # TODO(takgi) For its implementation reason, cupy.ndarray.argsort
-    # currently performs full argsort with Thrust's efficient radix sort
-    # algorithm.
+    # For simplicity, max_k is round up to the power of 2. If max_k is
+    # already the power of 2, it is round up to the next power of 2 because
+    # we need to collect the first max(kth)+1 elements.
+    max_k = max(32, 1 << max_k.bit_length())
 
-    # kth is ignored.
-    return data.argsort(_axis)
+    # The parameter t is the length of the list that stores elements to be
+    # selected for each thread. We divide the array into sz subarrays.
+    # These parameters are determined from the measurement on TITAN X.
+    t = 4
+    sz = 512
+    while sz > 0 and length // sz < max_k + 32 * t:
+        sz //= 2
+    sz *= self.size // length
+
+    # If the array size is small or k is large, we simply sort the array.
+    if length < 32 or sz < 1 or max_k >= 1024:
+        # kth is ignored.
+        return data.argsort(axis=-1)
+
+    shape = data.shape
+    data = data.ravel()
+    indices = cupy.arange(0, data.shape[0], dtype=cupy.int64)
+
+    # For each subarray, we collect first k elements to the head.
+    kern, merge_kern = _argpartition_kernel(self.dtype)
+    block_size = 32
+    grid_size = sz
+    kern(grid=(grid_size,), block=(block_size,), args=(
+        data, indices, max_k, self.size, t, sz))
+
+    # Merge heads of subarrays.
+    s = 1
+    while s < sz // (self.size // length):
+        block_size = 32
+        grid_size = sz // s // 2
+        merge_kern(grid=(grid_size,), block=(block_size,), args=(
+            data, indices, max_k, self.size, sz, s))
+        s *= 2
+
+    return indices
```

Comment:
I'm not sure why the indices returned by the `argpartition` kernel yield a different result from `partition` when they are used against the original input, this might explain why this implementation does not work for arrays with `ndim > 1`. Any pointers here might be useful

Reply:
Let me see it further, may take a while.

Reply:
The result of `partition` is written back to the original array:
https://github.com/cupy/cupy/blob/75fbfe2d0e7632dcdafef17ce5856b19e165d135/cupy/_core/_routines_sorting.pyx#L199

, so we can see the result with the shape of `self` (please remember `partition` is an in-place operation).

Conversely, we need to reshape `indices` to make it have the same shape with the input.